environments/common/src/main/java/org/jboss/weld/environment/deployment/discovery/ClassPathBeanArchiveScanner.java

```diff
@@ -0,0 +1,149 @@
+
+    private static final String BEANS_XML_FOUND_MESSAGE = "beans.xml found in {0}";
+
+    private static final String BEANS_XML_NOT_FOUND_MESSAGE = "beans.xml not found in {0}";
+
+    /**
+     *
+     * @param bootstrap
+     */
+    public ClassPathBeanArchiveScanner(Bootstrap bootstrap) {
+        super(bootstrap);
+    }
+
+    @Override
+    public List<ScanResult> scan() {
+        String javaClassPath = AccessController.doPrivileged(new GetSystemPropertyAction(JAVA_CLASS_PATH_SYSTEM_PROPERTY));
+        if (javaClassPath == null) {
+            throw CommonLogger.LOG.cannotReadJavaClassPathSystemProperty();
+        }
+        ImmutableList.Builder<ScanResult> results = ImmutableList.builder();
+        Set<String> entries = ImmutableSet.of(javaClassPath.split(Pattern.quote(File.pathSeparator)));
+        logger.debugv("Scanning class path entries: {0}", entries);
+        for (String entry : entries) {
+            if (entry == null || entry.isEmpty()) {
+                continue;
+            }
+            File entryFile = new File(entry);
+            if (!entryFile.canRead()) {
+                throw CommonLogger.LOG.cannotReadClassPathEntry(entryFile);
+            }
+            try {
+                if (entryFile.isDirectory()) {
+                    scanDirectory(entryFile, results);
+                } else {
+                    scanJarFile(entryFile, results);
+                }
+            } catch (IOException e) {
+                throw CommonLogger.LOG.cannotScanClassPathEntry(entryFile, e);
+            }
+        }
+        return results.build();
+    }
+
+    private void scanDirectory(File entryDirectory, ImmutableList.Builder<ScanResult> results) throws IOException {
+        // First try to find beans.xml
+        File beansXmlFile = new File(entryDirectory, AbstractWeldDeployment.BEANS_XML);
+        if (beansXmlFile.canRead()) {
+            logger.debugv(BEANS_XML_FOUND_MESSAGE, entryDirectory);
+            final BeansXml beansXml = parseBeansXml(beansXmlFile.toURI().toURL());
+            if (accept(beansXml)) {
+                results.add(new ScanResult(beansXml, entryDirectory.getPath()));
+            }
+        } else {
+            // No beans.xml found - check whether the bean archive contains an extension
+            logger.debugv(BEANS_XML_NOT_FOUND_MESSAGE, entryDirectory);
+            File extensionFile = new File(entryDirectory, EXTENSION_FILE);
+            if (!extensionFile.canRead()) {
+                results.add(new ScanResult(null, entryDirectory.getPath()));
+            }
+        }
+    }
+
+    private void scanJarFile(File entryFile, ImmutableList.Builder<ScanResult> results) throws IOException {
+        try (ZipFile zip = new ZipFile(entryFile)) {
+            Enumeration<? extends ZipEntry> entries = zip.entries();
+            BeansXml beansXml = null;
+            boolean extensionFound = false;
+            while (entries.hasMoreElements()) {
```

Comment:
Unless I am missing something it would be likely better to use this method: https://docs.oracle.com/javase/8/docs/api/java/util/zip/ZipFile.html#getEntry-java.lang.String-
instead of iterating over all the entries here.